atc/engine/build_step_delegate.go

```diff
@@ -225,139 +211,71 @@ func (delegate *buildStepDelegate) Errored(logger lager.Logger, message string)
-
 func (delegate *buildStepDelegate) FetchImage(
 	ctx context.Context,
-	image atc.ImageResource,
-	types atc.VersionedResourceTypes,
+	getPlan atc.Plan,
+	checkPlan *atc.Plan,
 	privileged bool,
-) (worker.ImageSpec, error) {
-	err := delegate.checkImagePolicy(image, privileged)
+) (worker.ImageSpec, db.ResourceCache, error) {
+	err := delegate.checkImagePolicy(getPlan.Get.Source, getPlan.Get.Type, privileged)
 	if err != nil {
-		return worker.ImageSpec{}, err
+		return worker.ImageSpec{}, nil, err
 	}
 
 	fetchState := delegate.state.NewLocalScope()
 
-	imageName := defaultImageName
-	if image.Name != "" {
-		imageName = image.Name
-	}
-
-	version := image.Version
-	if version == nil {
-		checkID := delegate.planID + "/image-check"
-
-		checkPlan := atc.Plan{
-			ID: checkID,
-			Check: &atc.CheckPlan{
-				Name:   imageName,
-				Type:   image.Type,
-				Source: image.Source,
-
-				VersionedResourceTypes: types,
-
-				Tags: image.Tags,
-			},
-		}
-
-		err := delegate.build.SaveEvent(event.ImageCheck{
-			Time: delegate.clock.Now().Unix(),
-			Origin: event.Origin{
-				ID: event.OriginID(delegate.planID),
-			},
-			PublicPlan: checkPlan.Public(),
-		})
+	if checkPlan != nil {
+		ok, err := fetchState.Run(ctx, *checkPlan)
 		if err != nil {
-			return worker.ImageSpec{}, fmt.Errorf("save image check event: %w", err)
-		}
-
-		ok, err := fetchState.Run(ctx, checkPlan)
-		if err != nil {
-			return worker.ImageSpec{}, err
+			return worker.ImageSpec{}, nil, err
 		}
 
 		if !ok {
-			return worker.ImageSpec{}, fmt.Errorf("image check failed")
-		}
-
-		if !fetchState.Result(checkID, &version) {
-			return worker.ImageSpec{}, fmt.Errorf("check did not return a version")
+			return worker.ImageSpec{}, nil, fmt.Errorf("image check failed")
 		}
 	}
 
-	getID := delegate.planID + "/image-get"
-
-	getPlan := atc.Plan{
-		ID: getID,
-		Get: &atc.GetPlan{
-			Name:    imageName,
-			Type:    image.Type,
-			Source:  image.Source,
-			Version: &version,
-			Params:  image.Params,
-
-			VersionedResourceTypes: types,
-
-			Tags: image.Tags,
-		},
-	}
-
-	err = delegate.build.SaveEvent(event.ImageGet{
-		Time: delegate.clock.Now().Unix(),
-		Origin: event.Origin{
-			ID: event.OriginID(delegate.planID),
-		},
-		PublicPlan: getPlan.Public(),
-	})
-	if err != nil {
-		return worker.ImageSpec{}, fmt.Errorf("save image get event: %w", err)
-	}
-
 	ok, err := fetchState.Run(ctx, getPlan)
 	if err != nil {
-		return worker.ImageSpec{}, err
+		return worker.ImageSpec{}, nil, err
 	}
 
 	if !ok {
-		return worker.ImageSpec{}, fmt.Errorf("image fetching failed")
+		return worker.ImageSpec{}, nil, fmt.Errorf("image fetching failed")
 	}
 
-	var cache db.UsedResourceCache
-	if !fetchState.Result(getID, &cache) {
-		return worker.ImageSpec{}, fmt.Errorf("get did not return a cache")
+	var result exec.GetResult
+	if !fetchState.Result(getPlan.ID, &result) {
+		return worker.ImageSpec{}, nil, fmt.Errorf("get did not return a result")
```

Comment:
Haven't looked into it too much, but I seem to be hitting this case pretty frequently
<img width="499" alt="Screen Shot 2021-06-22 at 11 33 07 AM" src="https://user-images.githubusercontent.com/26583442/122954655-a048cc00-d34d-11eb-9d09-5b51d50566e7.png">